### PR TITLE
Allow backup of mat view with dropped inner table instead of failing the backup

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow backup of materialized view with dropped inner table instead of failing the backup.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 7883a74f6d1dccecbb7717b842b4624bddadd22a with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56387/7883a74f6d1dccecbb7717b842b4624bddadd22a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6774763393">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56387/7883a74f6d1dccecbb7717b842b4624bddadd22a/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+${CLICKHOUSE_CLIENT} -nm --query "
+backup table ${CLICKHOUSE_DATABASE}.\`mv\` on cluster test_shard_localhost to Disk('backups', '${CLICKHOUSE_TEST_UNIQUE_NAME}');
+" | grep -o "BACKUP_CREATED"
+
+${CLICKHOUSE_CLIENT} -nm --query "
+drop table mv;
+restore table ${CLICKHOUSE_DATABASE}.\`mv\` on cluster test_shard_localhost from Disk('backups', '${CLICKHOUSE_TEST_UNIQUE_NAME}');
+" | grep -o "RESTORED"
+
+${CLICKHOUSE_CLIENT} --query "select count() from mv;"
```
Comment (vitlibar) [inline tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh:28]: As far as I know `SELECT` from a materialized view with its internal table dropped should throw an exception, shouldn't it?

File: tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+drop table if exists mv;
+create materialized view mv (a Int32) engine = MergeTree() order by tuple() as select * from src;
+"
+
+uuid=$(${CLICKHOUSE_CLIENT} --query "select uuid from system.tables where table='mv' and database == currentDatabase()")
+inner_table=".inner_id.${uuid}"
+${CLICKHOUSE_CLIENT} -nm --query "drop table \`$inner_table\` sync"
+
+${CLICKHOUSE_CLIENT} -nm --query "
+backup table ${CLICKHOUSE_DATABASE}.\`mv\` on cluster test_shard_localhost to Disk('backups', '${CLICKHOUSE_TEST_UNIQUE_NAME}');
```
Comment (vitlibar) [inline tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh:20]: `on cluster test_shard_localhost` is not necessary here, and in the restore command below as well.

File: tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh (hunk 1 above)
Comment (kssenii) [reply tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh:28]: yes, but after restore it started to work again so I assume empty inner table was created as a result of a restore.

**Review by vitlibar — APPROVED** [review]

File: tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh (hunk 1 above)
Comment (vitlibar) [reply tests/queries/0_stateless/02907_backup_mv_with_no_inner_table.sh:28]: ok

**Comment by kssenii** [comment]
> Stateless tests (aarch64) — fail: 1

`+noisy Debug messages	0.12227843707866669`
I did not add any Debug level logging (only Warning level), so I assume this failure is not related to changes.

> Stateless tests (release, analyzer) — fail: 1

`01287_max_execution_speed`
Also unrelated.

**Comment by tavplubix** [comment]
> > Stateless tests (aarch64) — fail: 1
> 
> `+noisy Debug messages 0.12227843707866669` I did not add any Debug level logging (only Warning level), so I assume this failure is not related to changes.

You're right, it's not.

But let's try to debug this. There's some statistics about log messages at the end of run.log, in particular:
```
2023-11-08 21:55:39 Top messages by level:
...
2023-11-08 21:55:39 (0.12214330063896353,'Not executing log entry {} of type {} for part {} because merges and mutations are cancelled now.')	Debug
```

Well, we could just add `LogFrequencyLimiter` to reduce the number of this garbage in log. Also it would be nice to add some backoff for replication queue entries.

But let's find tables that log this:
```
rg -Faz "because merges and mutations are cancelled now" clickhouse-server.log.zst | grep -Eo "<Debug> .*\(" | sort | uniq -c
...
      2 <Debug> test_qwoa6tcw.alter_table_4 (
 493169 <Debug> test_qwoa6tcw.alter_table_6 (
 485487 <Debug> test_qwoa6tcw.alter_table_8 (
 468267 <Debug> test_qwoa6tcw.alter_table_9 (
      5 <Debug> test_uzu7pkwj.replicated_table_for_mutations (
      9 <Debug> test_voa5tkzr.table_for_rename_replicated (
    272 <Debug> test_x50uyamy.concurrent_optimize_table (
...
```

Let's find the test:
```
rg -Faz "test_qwoa6tcw" clickhouse-server.log.zst| head -1
2023.11.08 21:51:11.498842 [ 211015 ] {} <Trace> DynamicQueryHandler: Request URI: /?query=CREATE+DATABASE+IF+NOT+EXISTS+test_qwoa6tcw&database=system&connect_timeout=30&receive_timeout=30&send_timeout=30&http_connection_timeout=30&http_receive_timeout=30&http_send_timeout=30&output_format_parallel_formatting=0&log_comment=00993_system_parts_race_condition_drop_zookeeper.sh
```

So it's `00993_system_parts_race_condition_drop_zookeeper`. Not surprising at all. But there are no `SYSTEM STOP MERGES` in this test, so it's not expected. Let's find when it started and ended:
```
rg -Faz "test_qwoa6tcw.alter_table_6" clickhouse-server.log.zst| grep -Fai cancel | head -1
2023.11.08 21:51:14.470062 [ 745 ] {} <Debug> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreeQueue): Not executing log entry queue-0000000134 of type MERGE_PARTS for part -4_0_5_1 because merges and mutations are cancelled now.
rg -Faz "test_qwoa6tcw.alter_table_6" clickhouse-server.log.zst| grep -Fai cancel | tail -1
2023.11.08 21:52:51.300858 [ 749 ] {} <Debug> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreeQueue): Not executing log entry queue-0000004113 of type MUTATE_PART for part 9_200_200_0_214 because merges and mutations are cancelled now.
```

Let's try to grep random substrings that may be interesting:
```
rg -Faz "test_qwoa6tcw.alter_table_6" clickhouse-server.log.zst| grep -Fai stop | head
2023.11.08 21:51:14.445133 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Information> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Stopped being leader
```

Wow, we were lucky, that's interesting. Let's find the query:
```
rg -Faz "3eda51bc-430d-49fd-ae20-8301635485f2" clickhouse-server.log.zst
10473165:2023.11.08 21:51:14.445044 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Debug> executeQuery: (from [::1]:45924) (comment: 00993_system_parts_race_condition_drop_zookeeper.sh) DROP TABLE IF EXISTS alter_table_6; (stage: Complete)
10473167:2023.11.08 21:51:14.445118 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> ContextAccess (default): Access granted: DROP TABLE ON test_qwoa6tcw.alter_table_6
10473173:2023.11.08 21:51:14.445133 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Information> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Stopped being leader
10473175:2023.11.08 21:51:14.445235 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreeRestartingThread): Restarting thread finished
10473179:2023.11.08 21:51:14.445246 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Will not wait for unique parts to be fetched by other replicas because shutdown called from DROP/DETACH query
10473216:2023.11.08 21:51:14.447784 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Waiting for threads to finish
12517273:2023.11.08 21:52:51.302303 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Threads finished
12517274:2023.11.08 21:52:51.302323 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (PartMovesBetweenShardsOrchestrator): PartMovesBetweenShardsOrchestrator thread finished
12517275:2023.11.08 21:52:51.302478 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Debug> DatabaseCatalog: Waiting for table 37081a09-2492-4bc7-b766-a972aca43b36 to be finally dropped
12521736:2023.11.08 21:52:51.840737 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Debug> TCPHandler: Processed in 97.395835416 sec.
```

So, the table has been dropped at `21:51:14`, but some background threads were unable to finish until `21:52:51`, it took 97 seconds to stop them. That's not okay. Looks like there's a lack of cancellation points somewhere. Can it take hours with production-size workloads?

**Comment by tavplubix** [comment]
Let's try to grep random substrings to find the culprit:
```
rg -Faz "test_qwoa6tcw.alter_table_6" clickhouse-server.log.zst| grep -Fai thread | head -100
...
2023.11.08 21:51:14.006679 [ 747 ] {} <Error> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Part -1_2_2_0 is lost forever.
2023.11.08 21:51:14.006728 [ 747 ] {} <Information> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Checking part 8_2_2_0
2023.11.08 21:51:14.008019 [ 747 ] {} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Part 8_2_2_0 in zookeeper: false, locally: false
2023.11.08 21:51:14.008034 [ 747 ] {} <Information> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Checking if anyone has a part 8_2_2_0 or covering part.
2023.11.08 21:51:14.022467 [ 747 ] {} <Error> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): No replica has part covering 8_2_2_0 and a merge is impossible: we didn't find smaller parts with either the same min block or the same max block.
2023.11.08 21:51:14.025041 [ 867 ] {} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Enqueueing -7_1_1_0 for check after after 0s
2023.11.08 21:51:14.290839 [ 848 ] {} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Enqueueing -1_3_3_0_4 for check after after 0s
2023.11.08 21:51:14.445235 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreeRestartingThread): Restarting thread finished
2023.11.08 21:51:14.447784 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Waiting for threads to finish
2023.11.08 21:51:14.498260 [ 679 ] {} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreeCleanupThread): Execution took 590 ms.
2023.11.08 21:52:51.241650 [ 747 ] {} <Warning> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Cannot create empty part 8_2_2_0 instead of lost. Will retry later
2023.11.08 21:52:51.241674 [ 747 ] {} <Trace> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Execution took 97234 ms.
2023.11.08 21:52:51.244037 [ 874 ] {} <Information> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Removing broken parts from ZooKeeper: 
2023.11.08 21:52:51.246068 [ 879 ] {} <Information> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Removing broken parts from ZooKeeper: 
2023.11.08 21:52:51.249354 [ 876 ] {} <Information> test_qwoa6tcw.alter_table_6 (ReplicatedMergeTreePartCheckThread): Removing broken parts from ZooKeeper: 
2023.11.08 21:52:51.302303 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (37081a09-2492-4bc7-b766-a972aca43b36): Threads finished
2023.11.08 21:52:51.302323 [ 283012 ] {3eda51bc-430d-49fd-ae20-8301635485f2} <Trace> test_qwoa6tcw.alter_table_6 (PartMovesBetweenShardsOrchestrator): PartMovesBetweenShardsOrchestrator thread finished
...
```

Wow, `Execution took 97234 ms`, that's interesting. Logs of the thread `[ 747 ]` around this time look like this: https://pastila.nl/?00110d92/6ac4ac002b96384419b992a61719dcf5#2vd0o3oy+jXVpQ9v4idcww==
The most interesting message is `Looks like log was updated or new replica appeared while creating new empty part, will retry`. It's an infinite loop in `StorageReplicatedMergeTree::createEmptyPartInsteadOfLost`.

**Comment by tavplubix** [comment]
Fix: https://github.com/ClickHouse/ClickHouse/pull/56533

**Pull request merged**